lib/galaxy/tool_util/verify/interactor.py

```diff
@@ -1511,6 +1509,10 @@ def register_exception(e: Exception):
 
     job_stdio = galaxy_interactor.get_job_stdio(job["id"])
 
+    if not (data_list or data_collection_list):
+        error = AssertionError("Tool produced no output datasets or collections")
+        register_exception(error)
```

Comment:
```suggestion
    assert data_list or data_collection_list, "Tool produced no output datasets or collections"
```

Reply:
We need to (at least it's the idea of the PR) record the assertion error to check for `expect_test_failure` at the end.

Reply:
I don't understand what you're saying. `expect_test_failure` is set by the test definition. What are you trying to fix ?

Reply:
We need to record the exception (i.e. call `register_exception` which stores in `found_exceptions`) in order to check for `expect_test_failure` https://github.com/galaxyproject/galaxy/blob/288a6c989a37be04a357191aeae9711f3b714806/lib/galaxy/tool_util/verify/interactor.py#L1625

If we just raise the Assertion then we have the same behavior as before. If we register it, we can react to this.